File: src/main/java/Library/holdList.java
package Library;

import java.util.*;

public class holdList {
    private final Map<Book, Queue<Borrower>> holds = new HashMap<>();

    public void placeHold(Borrower borrower, Book book){
        Queue<Borrower> queue = holds.get(book);
        if (queue == null){
            queue = new ArrayDeque<>();
            holds.put(book, queue);
        }
        queue.add(borrower);

    }

    // Check if books is available
    public List<Book> bookAvailable(Borrower borrower){
        List<Book> result = new ArrayList<>();
        for(Map.Entry<Book, Queue<Borrower>> entry: holds.entrySet()){
            Book book = entry.getKey();
            Queue<Borrower> queue = entry.getValue();

            if(book.getStatus() == Book.Status.Available &&
                    queue != null &&
                    !queue.isEmpty() &&
                    queue.peek().equals(borrower)){
                result.add(book);
            }
        }
        return result;
    }

    public boolean isOnHold(Book book, Borrower borrower) {
        Queue<Borrower> queue = holds.get(book);
        if (queue == null || queue.isEmpty()) {
            return false;
        }
        Borrower first = queue.peek();

        if (first.equals(borrower)) {
            return false;
        }
        return true;
    }

}

File: A3/models/holdList.js
// holdList.js

const Book = require("./Book");

class holdList {
    constructor() {
        // Map<Book, Queue<Borrower>>
        this.holds = new Map();
    }

    // placeHold(Borrower borrower, Book book)
    placeHold(borrower, book) {
        let queue = this.holds.get(book);
        if (!queue) {
            queue = []; // JS array used as FIFO queue
            this.holds.set(book, queue);
        }
        queue.push(borrower);
    }

    // bookAvailable(Borrower borrower)
    bookAvailable(borrower) {
        const result = [];

        for (const [book, queue] of this.holds.entries()) {
            if (
                book.getStatus() === Book.Status.On_hold &&
                queue &&
                queue.length > 0 &&
                queue[0] === borrower
            ) {
                result.push(book);
            }
        }

        return result;
    }

    // isOnHold(Book book, Borrower borrower)
    isOnHold(book, borrower) {
        const queue = this.holds.get(book);
        const first = queue ? queue[0] : null;

        if (first === borrower) {
            return false;
        }
        return true;
    }

    // hasReserved(Book book, Borrower borrower)
    hasReserved(book, borrower) {
        const queue = this.holds.get(book);
        if (!queue) return false;
        return queue.includes(borrower);
    }

    // checkReservation(Book book)
    checkReservation(book) {
        const queue = this.holds.get(book);
        return queue !== undefined && queue.length > 0;
    }

    // removeHold(Book book, Borrower borrower)
    removeHold(book, borrower) {
        const queue = this.holds.get(book);
        if (queue && queue.length > 0) {
            const index = queue.indexOf(borrower);
            if (index !== -1) {
                queue.splice(index, 1);
            }
        }
    }

    // getQueue(Book book)
    getQueue(book) {
        return this.holds.get(book) || [];
    }
}

module.exports = holdList;
